kythe/docs/rfc/2909.md

```diff
@@ -0,0 +1,145 @@
+| `"fd"-file-70-start-end-hash` | `Diagnostic` | `Diagnostic` per span (use `-1:-1` span for entire file) |
+
+
+Notes:
+
+* Each value will be contained within a wrapper message to support evolution of the table format
+* All file decoration entries are prefixed by "fd" and the file's `VName`
+* `VNames` will be stored instead of [Kythe tickets](https://kythe.io/docs/kythe-uri-spec.html)
+* The `FileDecorationsIndex` is used to store metadata for the rest of the entries and acts as an existence check
+* File text is chunked to support very large files (most files should be a single entry)
+* Decorations are ordered by span to support a narrow `FileDecorationsRequest` (group 10)
+* Node facts/definitions (groups 30 and 40) are stored after decorations and overrides to allow for the reader to skip over unneeded nodes (due to a narrow request)
+* Each definition is stored uniquely (group 50) and are related to nodes separately (group 40)
+
+## CrossReferences
+
+### Existing ProtocolBuffer definition:
+
+```protobuf
+// Simplified PagedCrossReference (no paging, removed deprecated features)
+message PagedCrossReferences {
+  string source_ticket = 1;
+  kythe.proto.common.MarkedSource marked_source = 6;
+  repeated string merge_with = 7;
+  
+  repeated Group group = 2;
+  Node source_node = 8;
+  
+  message Group {
+    string kind = 1;
+    repeated ExpandedAnchor anchor = 2;
+    repeated RelatedNode related_node = 3;
+    repeated Caller caller = 4;
+  }
+  message RelatedNode {
+    Node node = 1;
+    int32 ordinal = 2;
+  }
+  message Caller {
+    ExpandedAnchor caller = 1;
+    string semantic_caller = 2;
+    kythe.proto.common.MarkedSource marked_source = 3;
+    repeated ExpandedAnchor callsite = 4;
+  }
+}
+```
+
+### Key-value format:
+
+| key | value | kind |
+|---|---|---|
+| `"xr"-source` | `CrossReferencesIndex` |  |
+| `"xr"-source-00-kind-file-start-end` | `srvpb.ExpandedAnchor` | Regular cross-reference |
+| `"xr"-source-10-kind-ordinal-node` | `srvpb.Node` | Related nodes |
+| `"xr"-source-20-caller` | `srvpb.ExpandedAnchor`+`MarkedSource` | Definition for Caller |
+| `"xr"-source-20-caller-file-start-end` | `srvpb.ExpandedAnchor` | Callsite |
+
+
+Notes:
+
+* Each value will be contained within a wrapper message to support evolution of the table format
+* All cross-reference entries are prefixed by "xr" and the node's `VName`
+* `VNames` will be stored instead of [Kythe tickets](https://kythe.io/docs/kythe-uri-spec.html)
+* The `CrossReferencesIndex` is used to store metadata for the rest of the entries (e.g. counts, node facts, `merge_with`) and acts as an existence check
+* `Group`s are not used (each `Group` message is spread across multiple KV entries with a shared prefix)
+* `kind` fields in keys include a "priority" to achieve desired sorting order (this could alternatively be included in the group number)
+* After their `kind`, cross-references are ordered by file location
+
+## Beam Pipeline
```

Comment:
Is there a link to docs for this, that we could add here?

Reply:
Opened #2913